### ISSUE TITLE
when TransferManager goes out of scope, it continues transfers and callbacks

### ISSUE THREAD

**Issue by diablodale** [aws/aws-sdk-cpp#651]
As your team is considering the code update for https://github.com/aws/aws-sdk-cpp/issues/647 when TransferManager goes out of scope, I have an inquiry for your team to consider.

With SDK v1.1.42, I am experiencing zombie TransferManager. This zombie lives on with no customer code holding references to it. It continues to upload any files that were `UploadFile()`. It continues to call all three callbacks (2x progress, 1x status). The first behavior is odd and requires technical or human followup. The second behavior is more troublesome because the callbacks can be instance methods or lambda's accessing instance member variables...both of which were on an instance of a Class which was already destroyed.

I discovered this when I moved my TransferManager into a class member variable with a mutex around it. Anyone that needs the TransferManager calls a member function that will lock w/ the mutex and create/return the TransferManager. For example in some member function for uploading you can get the TransferManager with:  `auto transferManager(getTransferManager());`

When my class is destroyed, its destructor is called. Then the C++ runtime destroys the class member variables. The runtime decrements the refCount for TransferManager, but it is not fully destroyed. Some part of it lives on. After my Class is destroyed, I continue to see my debug console output showing a few calls to the status callback (inprogress, complete), many calls to the progress callback, and eventually the whole file is stored on S3 which I can see in the AWS console. I'm somewhat amazed there isn't an exception when TransferManager calls my callback lambda which accesses some instance member variables of a destroyed instance. Yes...I am already adjusting my code to use shared_ptr<> on the two variables at-risk.

Now that I see this behavior of TransferManager in action...it is unusual. Here is my thinking and inquiry for discussion:

* When I destroy a resource like TransferManager, I usually want it to halt and cleanup its children/activities. For example, I want TransferManager to `cancel()` all transfers from its internal queue.
* Naturally, need to consider how to handle multipart uploads on such a destroy. Does TransferManager also abort them? Or, is there a flag on TransferManager which indicates if the SDK customer wants such abort behavior? Also possible is that the `transferStatusUpdatedCallback` could react to TransferManager destructor iterating `cancel()` on its queue. That puts the SDK customer in control. I prefer this latter approach because I can code the multipart-abort once in the `transferStatusUpdatedCallback` and it would work both in the destroy and in any `cancel()` I do explicitly. I see a related comment on this at https://github.com/aws/aws-sdk-cpp/issues/259#issuecomment-243268280
* TransferManager has no interface to return the transfers in the transfer queue. Meaning, there is no iterator which allows me to easily do the following. Yes, I can code/maintain my own but this can be hard for all of Amazon's customers in bulk. Even the below requires a mutex and related locking code elsewhere that does down/uploads. Why? Because a separate thread could call `UploadFile()` while the destructor is iterating through the collection of `TransferHandle`s. My instinct is such an API would provide collection of shared_ptr<TransferHandle>. Then an SDK customer could do the following in a class destructor:
```cpp
lock_guard<mutex> lock(m_transferManagerMutex);
for (auto oneTransfer : myTransferManager->getTransfers()) {
    oneTransfer.cancel();
    oneTransfer.waitToFinish();
    if (hTransfer.IsMultipart()) {
        myTransferManager->AbortMultipartUpload(oneTransfer);
    }
}
```

**Comment by singku** [comment]
When TransferManager is done and ready to be destructed, it will wait for threads executor to cancel/abort all the in process tasks and destroy the threads executor owned by it, then to destroy itself. If that's not the case, there must be a bug.

We will take a look at it. Thank you for your report.

**Comment by marcomagdy** [comment]
@diablodale while working on the fix for issue #647 we've discovered a more serious problem. 

Here's the backstory in case you're interested.

Issue #574 reported that when TransferManager's life is shorter than one of its worker-threads, callbacks happen in the worker threads which invoke code in the-already-destroyed-TransferManager. This obviously resulted in a crash.
The fix for that bug was to simply ensure TransferManager's lifetime is extended as long as the worker-threads are still running. See ab8a76ec000f4a378cd0db4857fe1f28702593c8.

There're two problems with that patch:
1. First problem is what you reported in issue #647, which is the lifetime of TransferManager is not extended _early_ enough. Particularly, when we submit the work to the worker thread.
This is trivially fixable by doing the following in TransferManager.cpp in all areas which we submit work to a worker thread (sample below):
```patch
@@ -118,7 +119,8 @@ namespace Aws
             auto handle = Aws::MakeShared<TransferHandle>(CLASS_TAG, bucketName, keyName, createFileFn);
             handle->ApplyDownloadConfiguration(downloadConfig);

-            m_transferConfig.transferExecutor->Submit([this, createFileFn, handle] { DoDownload(handle); });
+            auto self = shared_from_this();
+            m_transferConfig.transferExecutor->Submit([self, createFileFn, handle] { self->DoDownload(handle); });
             return handle;
}
```

2. More importantly, this exposed a nasty lifetime dependency between `TransferManagerConfiguration::transferExecutor` and `TransferManager`. 
Basically, in the scenario which a worker thread destroys the last `shared_ptr` holding on to TransferManager, its destructor gets invoked, which in turn destroys the executor. When the executor is destroyed, it joins all its worker threads, this includes the thread which is currently running this code. A thread cannot join itself, so we crash.

We're working on the patch. Once it gets reviewed internally, we'll post PR externally to get more feedback before we merge it to master.

Hope this helps.

**Comment by diablodale** [comment]
Got it. Thanks for being in communication. In parallel, I have restructured my class to externally manage the lifetime of TransferManager and any TransferHandles representing transfers. In short, I have

```cpp
mutex                           m_s3ClientMutex;
shared_ptr<S3Client>            m_s3Client;    
mutex                           m_transferManagerMutex;
shared_ptr<TransferManager>     m_transferManager;
shared_ptr<mutex>               m_transfersMutex;
shared_ptr<unordered_set<shared_ptr<TransferHandle>>> m_transfers;
bool                            m_transfersShutdown;
```

* I use those for create, captures into lambdas, etc. Mutex around most of the things.
* I track all transfers with the `unordered_set`. Before `emplace()` in `up/downloadFile()` I check the boolean Shutdown flag.
* In `transferStatusUpdatedCallback` for a set of criteria on CANCELED/FAILED uploads, I call the synchronous `auto abortOutcome(s3Client->AbortMultipartUpload(abortRequest));`
* Also in `transferStatusUpdatedCallback` I check the Shutdown flag before I `unordered_set->erase()` for several (but not all) TransferStatus.
* My class's destructor `lock_guard(transfersMutex)`, sets the Shutdown flag, and for all `TransferHandles` in the `unordered_set` calls `cancel()`. 
* After the class is destroyed, there is some continued async activity which happens in `transferStatusUpdatedCallback` as all the async CANCEL events occur. Now in this lambda, those events are caught and I clean up any multipart uploads using the same synchronous abort() code in the 3rd bullet above
* Since the callback for all those cancel() on shutdown is via a lambda, they are running on an AWS SDK managed thread that is async/detached from my applications codebase. Given this, I don't mind using the synchronous abort call because it doesn't affect my application. If you have a devilish tester in your AWS team, I suggest they test the async abort in this kind of scenario to ensure that it will work.

**Comment by marcomagdy** [comment]
Once we patch master, destruction/cleanup will be more deterministic. TransferManager will join all the threads it spawned in its destructor.

**Comment by marcomagdy** [comment]
Take a look at PR https://github.com/aws/aws-sdk-cpp/pull/659 potentially build your application with that branch and see if you still experience the racy behavior.

**Comment by diablodale** [comment]
When I built this branch of the SDK with cmake, I got two compile errors
```
C:\repos\aws-sdk-cpp\aws-cpp-sdk-core\source\utils\threading\Executor.cpp(56): error C3861: 'assert': identifier not found
C:\repos\aws-sdk-cpp\aws-cpp-sdk-core\source\utils\threading\Executor.cpp(72): error C3861: 'assert': identifier not found
```

My editor VS2017 told me assert() is unknown. I had to add the following line to the top of `Executor.cpp`
```cpp
#include <cassert>
```
After that change, the SDK was able to be built with the following build settings:

```json
{
  "name": "x64-Debug",
  "generator": "Visual Studio 15 2017 Win64",
  "configurationType": "Debug",
  "inheritEnvironments": [ "msvc_x64" ],
  "buildRoot": "${env.USERPROFILE}\\CMakeBuilds\\${workspaceHash}\\build\\${name}",
  "cmakeCommandArgs": "-T v140",
  "buildCommandArgs": "-v:normal",
  "ctestCommandArgs": "",
  "variables": [
    {
      "name": "TARGET_ARCH",
      "value": "WINDOWS"
    },
    {
      "name": "CMAKE_INSTALL_BINDIR",
      "value": "${projectDir}/bin/"
    },
    {
      "name": "CMAKE_INSTALL_LIBDIR",
      "value": "${projectDir}/bin/"
    },
    {
      "name": "CMAKE_INSTALL_INCLUDEDIR",
      "value": "${projectDir}/bin/"
    },
    {
      "name": "BUILD_ONLY",
      "value": "s3;transfer"
    },
    {
      "name": "CUSTOM_MEMORY_MANAGEMENT",
      "value": "0"
    },
    {
      "name": "BUILD_SHARED_LIBS",
      "value": "0"
    },
    {
      "name": "FORCE_SHARED_CRT",
      "value": "1"
    },
    {
      "name": "ENABLE_TESTING",
      "value": "0"
    }
  ]
}
```

Please note, this is the first time I've attempted to locally compile the SDK. All my previous usage has been with the nuget packages of the SDK.

Next, I tried to update the sample app I provided in https://github.com/aws/aws-sdk-cpp/issues/647. I had the following issues:

* The API changed. OK. I'll adapt.
* I copied the boilerplate function Aws::UniquePtr<S3Client> MyS3::CreateS3Client() from your test app. Such won't work unchanged in my app because I need configs, s3clients, and transferManagers that are owned by independently instantiated classes based on the data within them. I can't make this factory non-static as I need `this` and its data to create the correct client config and, from it, the S3 client. I could pass in a copy of `this` and/or the client config to the factory as a parameter. This leads to changes in TransferManager or its config to take a copy of this. No. This is progressively messy and suggests to me there is a problem in this particular factory approach.
* This factor appears very much boilerplate to me. The only thing not is the client config. I believe this client config should be something passed into TransferManager either in constructor or as a property on the TransferManagerConfiguration. Probably the latter is best given the existing pattern of how an S3 client is created. However...this is weird and increasingly complex. This increasing complexity of these two bullets is caused by the following bullet point.
* TransferManagerConfiguration.s3ClientFactory as a dissociated exclusively owned `UniquePtr<S3Client>` is discouraging to me. A dup of `s3Client` resources. I assert() that a large amount of apps that would use `TransferManager` will also have an `s3Client` for listing, deletes, etc. To me this is a great opportunity for a `shared_ptr<s3Client>`. Then I (as the app dev) have a choice of passing my app's `shared_ptr<s3client>` to `TransferManager` --or-- I can `make_shared<s3client>` to give `TransferManager` one of its own. I like this flexibility. Today, I already have an s3Client "factory" function that delivers `shared_ptr<s3Client>`. I use it all over my app. I love it. _I recommend_ `TransferManagerConfiguration.s3ClientFactory` change to accept `shared_ptr<s3client>`. This will resolve all the above issues.
* Dozens of linking errors related to unfound external symbols for BCrypt, InternetOpen, InternetConnectA, HttpOpen, WinHttpConnect, GetUserProfileDirectoryW, GetFileVersionInfoSizeA, etc. This appear to be calls in aws-cpp-sdk-core.lib(CryptoImpl.obj), aws-cpp-sdk-core.lib(WinINetSyncHttpClient.obj), aws-cpp-sdk-core.lib(FileSystem.obj), aws-cpp-sdk-core.lib(OSVersionInfo.obj), etc.
 
I was not able to link the SDK into the sample app I provided in https://github.com/aws/aws-sdk-cpp/issues/647 without additional SDK-esque work. In my app's project, I had to add five system libraries for it to link in: `winhttp.lib, bcrypt.lib, wininet.lib, userenv.lib, version.lib`. I would prefer the SDK itself to have done this.

After adding these five libraries, my app compiled. :-) I tried to upload and got an assert on line 451 https://github.com/aws/aws-sdk-cpp/blob/57a5f9d83a5df9f866dd1f07f189caab71c1c308/aws-cpp-sdk-transfer/source/transfer/TransferManager.cpp#L451
```
Unhandled exception thrown: read access violation.
this->m_s3._Mypair.**_Myval2** was nullptr.
```

**Comment by marcomagdy** [comment]
The reason for S3Client being a unique_ptr owned by TransferManager is the essence of solving the race-conditions.

S3Client makes async calls within TransferManager. So, for us to deterministically join all the threads which were spawned by it, we have to be able to destroy it. If it's shared (via shared_ptr) then we have no control over that.

The factory is the most reasonable approach to have the user configure a S3Client instance and pass it to TransferManager. If you have a better idea on how to accomplish this, please let me know.

Finally, the assertion failure you mentioned brought to my attention that S3Client does not destroy its own thread executor in the destructor, because it's holding a shared_ptr to the thread executor. So, this needs to be fixed as well.

**Comment by diablodale** [comment]
To explore this, I suggest a step back/meta. What does an SDK user like me want to occur when a TransferManager is destroyed. Yes, technically at a low-level I want it to terminate/cleanup the resources/things it did when alive. Yet, at a high-level...what do I want?

I think what I want is for TransferManager to stop doing work. I want it to stop transferring everything.

This means that I want TransferManager to:
* allow no more transfers to be added to the internal transfer queue. It should ignore or error any attempts.
* no more traversing dir/folders with `UploadDir()` and `DownloadToDir()`
* `cancel()` any transfers that are in-progress. I want clean cancels. Why? Because this is a clean destruction of TM. This isn't an app crash (and this sdk doesn't do exceptions).
* don't make any more calls to the `s3Client` to get keys, e.g. due to a `DownloadToDir()`
* start no further transfers that were already in its internal transfer queue

Adding details to this:
* Normal TM callbacks like progress and status should be called as a result of all `cancel()` called in its destructor. This allows the user of the SDK to maintain to monitor/track these async operations. My code will know that it is causing TM to destruct. I can have functions on the TM callbacks to manage such destruction, e.g. aborting multipart uploads.
* typical memory releases/cleanup which should be automatically done through RIAA

In my thinking, `TransferManager` owns `TransferHandles`. Therefore TransferManager's job on destruction is to destroy TransferHandles and its internal queue holding them. Cascading that, a TransferHandle's job on destruction is to stop its own transfer which is done with a `cancel()`.

Much of the above thinking is driven by my current codebase. I have a clean working async `TransferManager` app for up/down/files/folders by wrapping TM with my own `unordered_set` to track `TransferHandles` and async monitoring of the same using all the current TM callbacks. I'm code complete on all that -- it works well. In my class' destructor, I do all the cancel() and unordered_set cleanup. If you think seeing my code would help you understand how a customer is using the SDK, I'm willing to privately send the C++ code to your corporate email address (for security/trust reasons).

What do you, at the very high-level, want TM to do when it is destroyed? 🤔

The high-level TransferManager I describe above, and my code running today, both don't need an exclusive `unique_ptr<s3Client>` to cleanly shutdown TM.  I do the clean "shutdown" in my wrapper class' destructor before RIAA destroys the TransferManager that my class owns. Since I can do it cleanly with a `shared_ptr<s3Client>` in my wrapper class, I am confident the SDK can do the same (probably better) in its internal code.

Some ideas:
* Reduce async activity due to the destructor of `TransferManager`. Perhaps do `cancel()` and then a blocking `WaitUntilFinished()` for each `TransferHandle`. It would make the TM destructor slower/block. However, it would prevent async `s3Client` activity after TM is destroyed. This feels similar to "join all the threads".
* Change the callback promise -> In the destructor of `TransferManager` when any `cancel()` is done, the code will not call any of the TM callbacks. This reduces async activity and the need to keep TM around to manage that callback. This idea puts the ownership on the SDK user (me) to do my own `cancel()` and callbacks before TM is destroyed if I want it clean w/ callbacks.

Regarding the client config... that a static function as a factory doesn't work for apps like mine that have more than one transfer manager and/or more than one client config... I can assign a lambda as the factory's function. Quick test suggests it works. And then use lambda captures of `shared_ptr<mutex>` and other such things to get my (potentially changing) client config into the factory so a correct `s3Client` can be created. This approach adds complexity, so it isn't preferred, though would probably work. **_A caution:_** I would _not_ want a single TransferManager to call and create more than one s3Client via the factory. Why? Because it is possible for my client config to change at any moment. Therefore, if TM created multiple s3Clients by calling the factory multiple times, it would be possible for one internally held client to have configA and another internally held client configB. That is definitely undesired.

**Comment by marcomagdy** [comment]
You bring up good points about what behavior to expect when shutting down. However, I don't think all the customers would prefer that behavior. So, I think it makes sense to add a `StopAllTransfers()` function to TransferManager.
By default, TransferManager will attempt to join all its worker threads in the destructor. 

I'm exploring other alternatives to the solution in PR #659.

Your feedback has been valuable. Thank you.

**Comment by marcomagdy** [comment]
We've added an enhancement to our backlog to add `StopAllTransfers` facility to TransferManager. I'm closing this bug for now.
Please reopen if you're still experiencing the same problem.
Thank you.

**Comment by kartikarcot** [comment]
This is still causing an issue. What is the right way to destroy a transfer manager? I tried destroying after I received one canceled event but that does not seem to work since the api might send multiple canceled notifications for one job and the callback would have captured variables that don't exist anymore.
